### PR TITLE
Option 2 - validate based on api yaml

### PR DESCRIPTION
**Pull request by kaitlyn-sharo**
Different from the first PR, this one has a mock yaml file for the api based on [this](https://nextcentury.atlassian.net/wiki/spaces/ITMC/pages/3041951763/Scenario+YAML+Documentation) 

The rest works approximately the same. We need to discuss which way is preferred moving forward.

Use validator2.py to test

### CODE REVIEW

**Review by brianpippin — APPROVED** [review]
Code looks good and I was able to execute and get these issues, and then was able to remove some by adding to the yaml file, well done.   This is a great start.  I will let Darren and or Larry chime in as well.

Warn (yamlValidator):	 'unstructured_postassess' is missing at the characters level of the yaml file
Warn (yamlValidator):	 'military_branch' is missing at the demographics level of the yaml file
Warn (yamlValidator):	 'rank' is missing at the demographics level of the yaml file
Warn (yamlValidator):	 'rank_title' is missing at the demographics level of the yaml file
Warn (yamlValidator):	 'skills' is missing at the demographics level of the yaml file
Warn (yamlValidator):	 'role' is missing at the demographics level of the yaml file
Warn (yamlValidator):	 'unstructured_postassess' is missing at the characters level of the yaml file
Warn (yamlValidator):	 'injuries' is missing at the characters level of the yaml file
Warn (yamlValidator):	 'unstructured_postassess' is missing at the characters level of the yaml file
Warn (yamlValidator):	 'injuries' is missing at the characters level of the yaml file
Warn (yamlValidator):	 'state' does not have a defined type at the scenes level of the yaml file
Warn (yamlValidator):	 'state' is missing at the scenes level of the yaml file
Warn (yamlValidator):	 'end_scenario_allowed' is missing at the scenes level of the yaml file
Warn (yamlValidator):	 'tagging' is missing at the scenes level of the yaml file
Warn (yamlValidator):	 'actions' is missing at the scenes level of the yaml file
Warn (yamlValidator):	 'transition_semantics' is missing at the scenes level of the yaml file
Warn (yamlValidator):	 'transitions' is missing at the scenes level of the yaml file
Warn (yamlValidator):	 'reference' should be type <class 'str'> but is <class 'int'> instead.
Warn (yamlValidator):	 'enabled' is missing at the tagging level of the yaml file
Warn (yamlValidator):	 'repeatable' is missing at the tagging level of the yaml file
Warn (yamlValidator):	 'probe_responses' is missing at the tagging level of the yaml file
Warn (yamlValidator):	 'unstructured' is not a valid key at the environment level of the yaml file. Allowed keys are dict_keys(['sim-environment', 'decision-environment'])
Warn (yamlValidator):	 'aid_delay' is not a valid key at the environment level of the yaml file. Allowed keys are dict_keys(['sim-environment', 'decision-environment'])
Warn (yamlValidator):	 'sim-environment' is missing at the environment level of the yaml file
Warn (yamlValidator):	 'decision-environment' is missing at the environment level of the yaml file
Warn (yamlValidator):	 'unstructured' is missing at the state level of the yaml file
Warn (yamlValidator):	 'mission' is missing at the state level of the yaml file
Warn (yamlValidator):	 'threat_state' is missing at the state level of the yaml file
Warn (yamlValidator):	 'supplies' is missing at the state level of the yaml file
Warn (yamlValidator):	 'characters' is missing at the state level of the yaml file
Warn (yamlValidator):	 'repeatable' is missing at the actions level of the yaml file
Warn (yamlValidator):	 'parameters' is missing at the actions level of the yaml file
Warn (yamlValidator):	 'condition_semantics' is missing at the actions level of the yaml file
Warn (yamlValidator):	 'conditions' is missing at the actions level of the yaml file
Warn (yamlValidator):	 'repeatable' is missing at the actions level of the yaml file
Warn (yamlValidator):	 'parameters' is missing at the actions level of the yaml file
Warn (yamlValidator):	 'condition_semantics' is missing at the actions level of the yaml file
Warn (yamlValidator):	 'conditions' is missing at the actions level of the yaml file
Warn (yamlValidator):	 'transition_semantics' is missing at the scenes level of the yaml file
Warn (yamlValidator):	 'transitions' is missing at the scenes level of the yaml file

Info (yamlValidator):	 test.yaml is not valid.

**Review by nextcen-dgemoets — COMMENTED** [review]
Fantastic work.  This is a great start!

It looks like you based your `api.yaml` on the yaml file in the `main` branch instead of the `development` branch.  I think it needs to be based on the [latest yaml](https://raw.githubusercontent.com/NextCenturyCorporation/itm-evaluation-server/development/swagger/swagger.yaml).  You'll also notice that in the latest yaml, we (well actually Kyle) moved the `enum`s into their own little model objects.  This helps us validate them at runtime (which, er, in theory we won't have to do if we assume yaml that passed the validator).

There are a lot fewer required fields than in the current `api.yaml` (I know you were just taking a guess at a lot of this because it's not specified yet).  But we can talk offline about what should be initially required.  It will of course evolve over time.

BTW, it occurs to me that we might need to use a different keyword for the `state` field in `Scenes` objects, because the idea is that you can specify any changes in state, but you don't have to specify things that haven't changed.  Actually, this will still be a problem for partially specified objects deeper in the state.  Hmm.  The alternative is to require complete state specification in each scene.  Anyway, not related to this ticket, but will impact the yaml.

We'll need to think about how to implement additional rules such as, "if `foo` is specified then `bar` must also be specified.  Or "if the value of `foo` is "bar", then the value of `squee` must be either `cat` or `dog`.  Anyway, I guarantee they're going to want complicated validation rules...

**Comment by nextcen-dgemoets** [comment]
It appears that the code requires there to be at least one `required` field, but I don't believe this will be the case.  E.g., `tagging`, where either `reference` is required or the other three fields are required.  Also, `condition`:  no one type of condition is required.

File: validator2.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+        '''
+        Checks the basic data types, allowing integers in place of floats
+        '''
+        return isinstance(item, type) or (type == float and isinstance(item, int))
+
+    def log_wrong_type(self, key, level, expected, actual):
+        '''
+        Logs when an incorrect type is found for a key
+        '''
+        self.logger.log(LogLevel.SEVERE_WARN, "Key '" + key + "' at level " + level + " should be type " + expected + " but is " + str(actual) + " instead.")
```
[the remark this answers is not in the document]
Comment (kaitlyn-sharo) [reply validator2.py]: Done, good catch

File: validator2.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+                    for x in location:
+                        ref_loc = ref_loc[x]
+                    if not self.validate_one_level(key, to_validate[key], ref_loc['properties'], ref_loc['required']):
+                        is_valid = False
+            found_keys.append(key)
+        # check for missing keys
+        for key in typed_keys:
+            if key not in found_keys:
+                if (key in required):
+                    self.logger.log(LogLevel.MINOR_WARN, "Required key '" + key + "' is missing at the " + level_name + " level of the yaml file")
```
[the remark this answers is not in the document]
Comment (kaitlyn-sharo) [reply validator2.py]: Got it

File: test.yaml [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+      rapport: 9 # A measure of closeness or affinity towards the player/medic. 0 represents strong dislike, 10 represents very close relationships like family.
+      demographics:
+        age: 34
+        sex: M
+        race: White # controlled vocab to be informed by research, TA4.
+        military_disposition: Allied US # controlled vocab includes Allied US, Allied, Civilian, Military Adversary, Non-Military Adversary
+        military_branch: US Marine Corps # controlled vocab includes US Army, US Navy, US Air Force, US Marine Corps, US Space Force, US Coast Guard
+        rank: E-4 # For controlled vocab, see "Paygrade" column of https://www.military.com/join-military/military-ranks-everything-you-need-know.html
+        rank_title: Corporal # For controlled vocab, see "Rank" column of https://www.military.com/join-military/military-ranks-everything-you-need-know.html
+        skills: # Describes abilities a character can have
```
[the remark this answers is not in the document]
Comment (nextcen-dgemoets) [reply test.yaml:89]: Yes, I see now.  That won't end up being required, but that explains why it gave the error-- thanks.
And yeah-- I thought of line numbers, and knew that would be tough.  I faced the same thing in the KAIROS validator.

**Comment by kaitlyn-sharo** [comment]
> It appears that the code requires there to be at least one `required` field, but I don't believe this will always be the case. E.g., `tagging`, where either `reference` is required or the other three fields are required. Also, `condition`: no one type of condition is required.

I noticed this while I was coding but forgot to go back and fix it. Updated now.

**Comment by kaitlyn-sharo** [comment]
> Current issues:
> 
> * Getting a crash, I think because `Threat` has no `properties`, only `additionalProperties`;
> * I think there's an issue with strings that look like integers (see `kdma_associations` in `actions` keyword); and
> * Need to check floats-- I updated representation in YAML and tried to address code, but the code changes might be incomplete.
> 
> Things that don't need to be addressed in this PR:
> 
> * `rank` and `rank_title` need correct controlled vocabulary;
> * Need to address how we want to be able to specify any changes in state at the scene level, but not have to specify things that haven't changed;
> * Adding numeric value ranges to YAML and validating in tool;
> * Need to figure out if initial state will be in `state` property at top level, or it is specified in the 0th scene; and
> * Incorporating all changes from `swagger.yaml` in the `development` branch.  Could do here or in another ticket.

Bugs are fixed and yes, you changed 'float' to 'number' in primitive types. That broke things. We can keep both, but the api could just specify ints vs floats. It doesn't bother me to keep it as a number. A string is validated now if the content is an int or float, and I did some extra checking for the additionalProperties object types and enums, since those formats in the api yaml broke things. There's probably some duplicated code and a few counting things and printing things and commenting things to clean up, so I'll start with that on Monday just to tighten it up and prepare for the next task.

**Comment by nextcen-dgemoets** [comment]
> Bugs are fixed and yes, you changed 'float' to 'number' in primitive types. That broke things. We can keep both, but the api could just specify ints vs floats. It doesn't bother me to keep it as a number. A string is validated now if the content is an int or float, and I did some extra checking for the additionalProperties object types and enums, since those formats in the api yaml broke things. There's probably some duplicated code and a few counting things and printing things and commenting things to clean up, so I'll start with that on Monday just to tighten it up and prepare for the next task. I also plan to finish making all of the enums into their own separate section. I wonder how you feel about keeping enums separate from the schemas?

I changed `float` to `number` because `float` isn't a valid type in swagger (as reported by https://editor.swagger.io/).  I knew there were implications in the code that I hadn't addressed-- thank you for addressing them.  :)  Although I think we can remove `'float': float` from the `PRIMITIVE_TYPE_MAP`.

The `kdma_association` property was copied from TA1, and they had them as strings.  That said, a number/float seems to be better so I guess we can change them.

Keeping the enums separate from the schemas was necessary to validate controlled vocabulary in the server, but maybe this is no longer necessary if there's a validator tool, right?  It creates quite a proliferation of model objects / code.  These enum objects *can* also be used by TA2 clients, however, to ensure that they're sending good values at runtime.  It's something we can discuss with Brian maybe.

**Review by nextcen-dgemoets — APPROVED** [review]
Looks really great.  I couldn't find any bugs.

The only thing that I *might* have expected the validator to do is detect duplicate keys-- if in a given level you specify the same key twice.  It currently silently ignores the first value.  This may be a function of the yaml package and not the validator.  Anyway, it could be another ticket for someday.

The other thing we discussed which is specifically out of scope for this ticket is to configure minimum and/or maximum values in the yaml and validate them in the tool.

**Comment by kaitlyn-sharo** [comment]
> Looks really great. I couldn't find any bugs.
> 
> The only thing that I _might_ have expected the validator to do is detect duplicate keys-- if in a given level you specify the same key twice. It currently silently ignores the first value. This may be a function of the yaml package and not the validator. Anyway, it could be another ticket for someday.
> 
> The other thing we discussed which is specifically out of scope for this ticket is to configure minimum and/or maximum values in the yaml and validate them in the tool.

That's a great point. I actually used to have that (maybe only in option 1?). It's super easy to add back in, I will do it now.

**Pull request merged**